Replace the previous ADAPTIVE coordinate with the flux-based AG implementation

angus-g opened the pull request:
Moving #6 to come from this fork so it can be deployed. Original description below, otherwise see the original PR for some discussion.

---

This changes the old `ADAPTIVE` coordinate that blended between grids with a new flux-based implementation called AG (**A**daptive **G**rid). In its current state, there are a few things that may need to be changed:

- `dt` gets passed into regridding
- There's an extra diagnostic registration stage so we can get information from *within* regridding (for debugging the adaptivity)
- T/S group pass halos have to grow

Due to the use of a stencil for determining the smoothing/adaptivity flux, there's a non-reproducibility between symmetric and non-symmetric memory modes that I haven't yet worked out. This is raised in the `tc2.b` case that I added for regression testing in the GitHub Actions suite.

This has been a part-time endeavour for ages, so I don't completely remember all the implementation specifics. Please criticise as necessary!

I will update with some of the mathematical justification, and a parameter description, which should go into the documentation.

angus-g committed:
Add flux-based AG coordinate

This is the adaptive coordinate from the Gibson PhD thesis. It
involves a flux-based calculation of density and smoothing adaptivity
terms, which are weighted depending on the local isopycnal
slope. Because these two terms alone aren't a sufficient constraint,
interfaces can be nudged toward a pre-defined grid using the
filtered_grid_motion mechanism. Additionally, to prevent convective
instabilities resulting purely from grid motion, water masses can be
"pushed" through vanished layers.

As a structural change, the filtered_grid_motion subroutine and its
controlling parameters are lifted out to the new ALE/filter_utils
module, so that the main AG code can reside in ALE/coord_adapt without
a circular dependency back on MOM_regridding.

angus-g committed:
Instrument the AG coordinate regridding with diagnostics

We can pull out the contributing terms in the coordinate from within
the regridding routine, and the framework is set up to pull arbitrary
information out from the regridding step. This should be useful for
coordinate debugging.

angus-g committed:
Bump up halos in tracer group pass

This probably needs slightly more careful analysis to make sure we're
really doing the right thing.

angus-g committed:
Add AG to DOME2d

angus-g committed:
Fix u/v loops in AG

angus-g committed:
Initialize dz_s and dz_p to 0 to avoid uninitialised diagnostics

These variables are only set on land points, but the diagnostics are
including them regardless. While it shouldn't affect the regular
operation of the code, the diagnostics could end up containing junk
data.

angus-g committed:
Add T scaling to AG parameters

The two timescales (adaptivity diffusivity and restoring) needed to be
annotated with the s_to_T scaling in order to pass the dimensional
consistency test for T.

angus-g committed:
Reorder build_grid_adapt params

We need to pass through the US structure.

angus-g committed:
Use epsilon for zero-denominator checks

We were failing the grid rotation test because of some floating point
differences. It makes much more sense to compare to epsilon anyway.

angus-g committed:
Factor out calculation of hd{i,j}_sig and hd{i,j}_sig_phys

There was a lot of duplicated code between these sections, making it
very hard to verify they were even written exactly the same.

angus-g committed:
Fix style errors to satisfy Actions

angus-g committed:
Add L and H scaling to AG

angus-g committed:
add tc2.b case for AG coordinate

angus-g committed:
Use velocity point mask for AG derivatives

Seems like we were bringing in junk data for slopes which was causing
differences across restarts.

angus-g committed:
Remove unused remapCS from build_grid_adaptive

angus-g committed:
Remove diag_cs requirement for AG

It's useful to be able to dump the diagnostics out, but we need this
weird "attach the diag CS later" method, which we might not always
want to do. Instead, let's just make diagnostics conditional.

angus-g committed:
Add OpenMP acceleration to build_adapt_grid

angus-g committed:
Explicitly set do_diag for adapt coord

angus-g committed:
Implement a 7-point horizontal filter for adapt weights

This filter only acts on points, if their masks are valid. Shouldn't
require any halo extension.

angus-g committed:
Ignore slope if density gradient inverts

angus-g committed:
Fix adapt weight filter calculation

A couple of issues here. First, there was a typo in the V-direction
filter, so it was using the wrong mask. More importantly, the stencil
was reaching through uncalculated cells and past the edges of the
domain, leading to FPE.

angus-g committed:
Add a group pass to fix horizontal filter consistency

angus-g committed:
Add filter_utils to CMakeLists

dougiesquire commented:
@angus-g is this ready for review from your perspective? My impressions from the discussion [in the previous PR](https://github.com/ACCESS-NRI/MOM6/pull/6#issuecomment-2731383720) is that there are still changes to come but let know if I've misunderstood

angus-g commented:
Thanks @dougiesquire. I think it's more or less ready from a broad-scale perspective. I guess there is a segfault in the grid test that will need to be addressed. ~~As picked up in that comment, I'll revert `tc2.b` and create a `tc2.c` that exercises AG instead.~~ `tc2.b` already is new... The regression test fails because the coordinate fundamentally changed.

 But I think most things to make the CI happier aren't likely to affect the design and implementation decisions broadly.

angus-g committed:
Fix openMP build

angus-g committed:
Fix indentation

angus-g committed:
Widen initial halo for density deriv

With non-symmetric memory, interpolation of hdi_sig from a U point
onto a V point accesses a non-initialised element because the halo
isn't wide enough. By consuming an initial 3-point halo of T/S, we
have enough data to perform the interpolation in both memory modes
